src/Common/SLRUCachePolicy.h

```diff
@@ -0,0 +1,246 @@
+        {
+            cell.is_protected = true;
+            current_protected_size += cell.size;
+            protected_queue.splice(protected_queue.end(), probationary_queue, cell.queue_iterator);
+        }
+
+        removeOverflow(protected_queue, max_protected_size, current_protected_size, /*is_protected=*/true);
+
+        return cell.value;
+    }
+
+    void set(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) override
+    {
+        auto [it, inserted] = cells.emplace(std::piecewise_construct,
+            std::forward_as_tuple(key),
+            std::forward_as_tuple());
+
+        Cell & cell = it->second;
+
+        if (inserted)
+        {
+            try
+            {
+                cell.queue_iterator = probationary_queue.insert(probationary_queue.end(), key);
+            }
+            catch (...)
+            {
+                cells.erase(it);
+                throw;
+            }
+        }
+        else
+        {
+            current_size -= cell.size;
+            if (cell.is_protected)
+            {
+                current_protected_size -= cell.size;
+                protected_queue.splice(protected_queue.end(), protected_queue, cell.queue_iterator);
+            }
+            else
+            {
+                cell.is_protected = true;
+                protected_queue.splice(protected_queue.end(), probationary_queue, cell.queue_iterator);
+            }
+        }
+
+        cell.value = mapped;
+        cell.size = cell.value ? weight_function(*cell.value) : 0;
+        current_size += cell.size;
+        current_protected_size += cell.is_protected ? cell.size : 0;
+
+        removeOverflow(protected_queue, max_protected_size, current_protected_size, /*is_protected=*/true);
+        removeOverflow(probationary_queue, max_size, current_size, /*is_protected=*/false);
+    }
+
+protected:
+    using SLRUQueue = std::list<Key>;
+    using SLRUQueueIterator = typename SLRUQueue::iterator;
+
+    SLRUQueue probationary_queue;
+    SLRUQueue protected_queue;
+
+    struct Cell
+    {
+        bool is_protected;
+        MappedPtr value;
+        size_t size;
+        SLRUQueueIterator queue_iterator;
+    };
+
+    using Cells = std::unordered_map<Key, Cell, HashFunction>;
+
+    Cells cells;
+
+    size_t current_protected_size = 0;
+    size_t current_size = 0;
+    const size_t max_protected_size;
+    const size_t max_size;
+    const size_t max_elements_size;
+
+    WeightFunction weight_function;
+
+    void removeOverflow(SLRUQueue & queue, const size_t max_weight_size, size_t & current_weight_size, bool is_protected)
+    {
+        size_t current_weight_lost = 0;
+        size_t queue_size = queue.size();
+
+        auto need_remove = [&]() -> bool
+        {
+            if (is_protected)
+            {
+                return ((max_elements_size != 0 && cells.size() - probationary_queue.size() > max_elements_size - probationary_queue.size())
```

Comment:
> it can be simplified: `cells.size() - probationary_queue.size() > max_elements_size - probationary_queue.size()` --> `cells.size() > max_elements_size`

There should be  `cells.size() - probationary_queue.size() > max_elements_size - probationary_queue.size()` --> `cells.size()  - probationary_queue.size() > max_elements_size` to correctly work with protected elements. These elements only move from protected to probationary part and total amount of elements doesn't change, so we should check if it's enough to remove all probationary elements after cleaning protected part to have less than max elements in cache